shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala

```diff
@@ -987,14 +978,27 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
             }
           }
           totalCountLoadedSoFar += pages.getRowCount
+        } catch {
+          case e => {
+            parquetFileReader.close()
+            throw e
+          }
         }
+      }
 
-        def nextBatch: Boolean = {
+      /**
+       * Read the next RowGroup and read each column and return the columnarBatch
+       */
+      def nextBatch: Boolean = {
+        try {
           for (vector <- columnVectors) {
             vector.reset()
           }
           columnarBatch.setNumRows(0)
-          if (rowsReturned >= totalRowCount) return false
+          if (rowsReturned >= totalRowCount) {
+            parquetFileReader.close()
```

Comment:
This has the problem that unless the iterator is drained completely, `parquetFileReader` is not guaranteed to be closed.

Instead of peppering `catch` clauses everywhere, I suggest making `CurrentBatchIterator` closeable and then install a task completion handler at the higher level that guarantees the iterator is always closed when the task completes, closing any other instances as you move between instances of the iterator along the way.

Reply:
OK, but then we still need to handle the case when/if an exception is thrown anywhere in `CurrentBatchIterator` don't we? May be we need to wrap this in the `CloseableIterator` in addition to this "peppering"? Or does the task completion handler take care of this even if there is an exception?

Reply:
No, we're basically delegating to the user of the iterator to ensure it's closed even in exceptions.  Think of it as the iterator sort of wrapping the parquet reader, the reader will be closed either when the iterator is drained or when the caller closes the iterator, and it's the responsibility of the iterator user to ensure it's closed even in exceptional conditions.

> Or does the task completion handler take care of this even if there is an exception?

Exactly.  The exception will bubble up and kill the task, causing the completion handler to be invoked.  Or worse-case if someone up the call stack eats the exception, the task will eventually complete in some form and ensure this is not indefinitely leaked.